### PR TITLE
adding only fifo.rs

### PR DESCRIPTION
**Pull request by nebelgrau77**
OK, trying to do it this way: this follows your instructions, that is only a single file `fifo.rs` was added to this branch. Let's see if it works.

### CODE REVIEW

File: src/fifo.rs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+impl FIFO_MODE {
+    pub fn value(self) -> u8 {
+        (self as u8) << 5 // shifted into the right position, can be used directly
+    }
+}
+
+/// Decimation of acceleration data on OUT REG and FIFO (Refer to table 65)
+#[allow(non_camel_case_types)]
+#[derive(Debug, Clone, Copy)]
+pub enum DECIMATE {
```
Comment (lonesometraveler) [inline src/fifo.rs:182]: Does `DECIMATE` have to be all uppercase? Can it be `Decimate`?

File: src/fifo.rs [hunk 2]
```diff
@@ -0,0 +1,197 @@
+//! Various functions related to FIFO
+//!
+//! TO DO:
+//! - CHECK IF ALL FUNCTIONS ARE IMPLEMENTED
+//! - MAKE SURE REGISTERS ARE NOT OVERWRITTEN BY MISTAKE
```
Comment (lonesometraveler) [inline src/fifo.rs:5]: Are these still TO DO?

**Review by lonesometraveler — CHANGES_REQUESTED** [review]
### LSM9DS1 methods
- `configure_fifo`
- `get_fifo_status`
- `set_decimation`

I am not sure if they should be in this file or in lib.rs. 


### FIFO struct/enum names
We have 
- `FIFO_Bitmasks`
- `FIFOConfig`
- `FifoStatus`
- `FIFO_MODE`

I want to make it consistent. I think `FifoSomething` or `FIFOSomething` works. What do you think?

**Review by lonesometraveler — CHANGES_REQUESTED** [review]

File: src/fifo.rs [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+            /// Is FIFO full and at least one sample has been overwritten?
+            fifo_overrun: match fifo_src & FIFO_Bitmasks::OVRN {
+                0 => false,
+                _ => true,
+            },
+            /// Is FIFO empty (no unread samples)?
+            fifo_empty: match fifo_level_value {
+                0 => true,
+                _ => false,
+            },
```
Comment (lonesometraveler) [inline src/fifo.rs:136]: Can we use matches! macro? Maybe it is more readable?

```rust
            /// Is FIFO filling equal or higher than the threshold?
            fifo_thresh_reached: matches!(fifo_src & FIFO_Bitmasks::FTH, 1),
            /// Is FIFO full and at least one sample has been overwritten?
            fifo_overrun: matches!(fifo_src & FIFO_Bitmasks::OVRN, 1),
            /// Is FIFO empty (no unread samples)?
            fifo_empty: matches!(fifo_level_value, 0),
```

File: src/fifo.rs (hunk 1 above)
Comment (nebelgrau77) [reply src/fifo.rs:182]: Good question. I see now that I wasn't consistent about it and some enums are ALL CAPS, while others are not. I tried checking here: [naming conventions](https://github.com/rust-lang/rfcs/blob/master/text/0430-finalizing-naming-conventions.md), but it only mentions enum variants specifically. But I guess enum is closer to types and traits, so it probably should be UpperCamelCase. If we agree on that, there's some more to change accross the code.

File: src/fifo.rs (hunk 2 above)
Comment (nebelgrau77) [reply src/fifo.rs:5]: It should be all OK, but I will check and then remove the comments.

File: src/fifo.rs [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    /// FIFO memory enable
+    pub fifo_enable: bool,
+    /// Select FIFO operation mode (see Table 84 for details)        
+    pub fifo_mode: FIFO_MODE, // default Bypass
+    /// Enable threshold level use
+    pub fifo_use_threshold: bool,
+    /// Set the threshold level
+    pub fifo_threshold: u8, // default 32
+    /// Store temperature data in FIFO
+    fifo_temperature_enable: bool,
```
[the remark this answers is not in the document]
Comment (nebelgrau77) [reply src/fifo.rs:38]: No, that's clearly a mistake, I'll fix it.

File: src/fifo.rs (hunk 3 above)
Comment (nebelgrau77) [reply src/fifo.rs:136]: Looks interesting, but I have never ever used it. I need to look it up :) If we decided to use it, it should be changed consistently accross the whole crate, right?

**Comment by nebelgrau77** [comment]
> LSM9DS1 methods
configure_fifo
get_fifo_status
set_decimation
I am not sure if these should be in this file or in lib.rs.

I would keep them in `fifo.rs`: they are strictly limited to FIFO operations, so it should be a separate module IMO. E.g. if someone is working on a similar LSMx driver and wants to add FIFO features, you have this whole block to plug&play. At least that was my "modular" concept here.   

> FIFO struct/enum names
I want to make it consistent. I think FifoSomething or FIFOSomething works. What do you think?

I would vote for FIFO_Something, but I see that it would be some bizarre "camel snake case" :D Following the guidelines I suppose it should be FIFOSomething (FIFOMode, FIFOConfig, etc.). Do you think it's still readable? I don't like "Fifo" as it distorts the idea of the acronym.

**Comment by lonesometraveler** [comment]
> > LSM9DS1 methods
> > configure_fifo
> > get_fifo_status
> > set_decimation
> > I am not sure if these should be in this file or in lib.rs.
> 
> I would keep them in `fifo.rs`: they are strictly limited to FIFO operations, so it should be a separate module IMO. E.g. if someone is working on a similar LSMx driver and wants to add FIFO features, you have this whole block to plug&play. At least that was my "modular" concept here.

I want to decouple modules. Right now, fifo.rs depends on outside modules. If we remove all the LSM9DS1 methods and delete `use super::*;`, it becomes truly modular. 
Imagine the situation where we want to rename enum `Sensor` for something else or add another parameter to `read_register()` (Unlikely for this particular method. But I hope you get the point). Do we really want to modify fifo.rs each time we make a change? 
A developer working on a similar LSMx driver cannot plug in your implementation of fifo.rs because it requires our `LSM9DS1`, `Sensor`, etc.. But if fifo.rs is decoupled from other modules, it could be plugged into another codebase.


> > FIFO struct/enum names
> > I want to make it consistent. I think FifoSomething or FIFOSomething works. What do you think?
> 
> I would vote for FIFO_Something, but I see that it would be some bizarre "camel snake case" :D Following the guidelines I suppose it should be FIFOSomething (FIFOMode, FIFOConfig, etc.). Do you think it's still readable? I don't like "Fifo" as it distorts the idea of the acronym.

`FifoSomething` for readability and `FIFOSomething` for accuracy? I like `FIFOSomething`.

File: src/fifo.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (nebelgrau77) [reply src/fifo.rs]: Done.

**Comment by nebelgrau77** [comment]
OK, I see your point about decoupling. I will do that in a second step, moving those functions to `lib.rs`. At this point I will also commit `lib.rs` together with `fifo.rs`, does that make sense?

File: src/fifo.rs (hunk 2 above)
Comment (nebelgrau77) [reply src/fifo.rs:5]: Checked everything, should be OK. Removed the comments.

File: src/fifo.rs (hunk 3 above)
Comment (nebelgrau77) [reply src/fifo.rs]: I'm not sure about this macro, as in: I don't really understand how it works (and so far found only not-very-clear docs), and I like the more explicit code much more. I don't think it is more readable, quite the contrary. I would leave it as is.

**Pull request merged**